<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Ensemble Classification using <code>splendid</code> • splendid</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">splendid</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AlineTalhouk/splendid">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Ensemble Classification using <code>splendid</code>
</h1>
            
          </div>

    
    
<div class="contents">
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Supervised learning is the branch of statistical learning where we construct a predictive model with the goal of regression or classification. The <code>splendid</code> package focuses on classification with an ensemble framework: many classification algorithms are used, and prediction is done across bootstrap replicates of the data. An ensemble classifier is built from the best performing algorithms according to evaluation measures. No single classifier always performs the best for every data set that exists, so there is increasing utility to come up with ensemble classifiers. The objective is to use this classifier to obtain highly accurate predictions in independent data sets for the purposes of diagnostic identification. In genomic studies for example, one might be interested in using the class labels of a cancer subtype in one cohort to predict the subtypes in an independent cohort.</p>
<p><code>splendid</code> is currently only available on GitHub (<em>the second line below will be uncommented once the repository becomes public</em>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages("devtools")</span>
<span class="co"># devtools::install_github("AlineTalhouk/splendid")</span>
<span class="kw">library</span>(splendid)
<span class="kw">library</span>(knitr)
<span class="kw">data</span>(hgsc)</code></pre></div>
</div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>The main function of <code>splendid</code> is <code><a href="../reference/splendid.html">splendid()</a></code>, and the usage is intuitive:</p>
<ul>
<li>
<code>data</code>: your rectangular data in standard form (rows as samples/observations, columns as variables/features)</li>
<li>
<code>class</code>: the true/reference class labels with length equal to the number of samples</li>
<li>
<code>n</code>: number of bootstrap replicates to generate as training sets. Test sets are taken from the out-of-bag samples (OBB).</li>
<li>
<code>seed</code>: constant for reproducibility in the bootstrap resampling</li>
<li>
<code>algorithms</code>: character vector of classification algorithms</li>
<li>
<code>rfe</code>: whether to use Recursive Feature Elimination (RFE) for feature selection. Only implemented for certain algorithms.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">class &lt;-<span class="st"> </span><span class="kw">attr</span>(hgsc, <span class="st">"class.true"</span>)
<span class="kw">table</span>(class)
<span class="co">#&gt; class</span>
<span class="co">#&gt; DIF.C4 IMM.C2 MES.C1 PRO.C5 </span>
<span class="co">#&gt;    135    107    109    138</span>
sl_result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splendid.html">splendid</a></span>(<span class="dt">data =</span> hgsc, <span class="dt">class =</span> class, <span class="dt">n =</span> <span class="dv">2</span>,
                      <span class="dt">algorithms =</span> <span class="kw">c</span>(<span class="st">"slda"</span>, <span class="st">"knn"</span>, <span class="st">"svm"</span>), <span class="dt">rfe =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>The resulting object is a list with the following elements:</p>
<ul>
<li>
<code>models</code> gives the model fits for each of the algorithms used. The fits are themselves nested based on the number of bootstrap replicates.</li>
<li>
<code>preds</code> has the same hierarchical structure as <code>models</code>, except the metadata consists of the predicted class labels.</li>
<li>
<code>evals</code> is a list of data frames showing evaluation metrics for each algorithm. Rows are metrics and columns are bootstrap replicates. More details on this element will be provided later.</li>
<li>
<code>bests</code> is a vector that shows the best-performing algorithm for each bootstrapped replicate of the data, chosen by rank aggregation.</li>
<li>
<code>ensemble_algs</code> tallies the algorithm frequencies in <code>bests</code>, returning the top (3, by default) algorithms chosen.</li>
<li>
<code>ensemble</code> is a list of model fits for each of the algorithms in <code>ensemble_algs</code>, fit on the full data.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(sl_result, <span class="dt">max.level =</span> <span class="dv">2</span>)
<span class="co">#&gt; List of 8</span>
<span class="co">#&gt;  $ models       :List of 3</span>
<span class="co">#&gt;   ..$ slda:List of 2</span>
<span class="co">#&gt;   ..$ knn :List of 2</span>
<span class="co">#&gt;   ..$ svm :List of 2</span>
<span class="co">#&gt;  $ preds        :List of 3</span>
<span class="co">#&gt;   ..$ slda:List of 2</span>
<span class="co">#&gt;   ..$ knn :List of 2</span>
<span class="co">#&gt;   ..$ svm :List of 2</span>
<span class="co">#&gt;  $ evals        :List of 3</span>
<span class="co">#&gt;   ..$ slda:'data.frame': 24 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ knn :'data.frame': 24 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ svm :'data.frame': 24 obs. of  2 variables:</span>
<span class="co">#&gt;  $ bests        : Named chr [1:2] "svm" "svm"</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:2] "1" "2"</span>
<span class="co">#&gt;  $ ensemble_algs: chr "svm"</span>
<span class="co">#&gt;  $ ensemble_mods:List of 1</span>
<span class="co">#&gt;   ..$ :List of 29</span>
<span class="co">#&gt;   .. ..- attr(*, "class")= chr "svm"</span>
<span class="co">#&gt;  $ seq_mods     : NULL</span>
<span class="co">#&gt;  $ seq_preds    : NULL</span></code></pre></div>
</div>
<div id="classification" class="section level2">
<h2 class="hasAnchor">
<a href="#classification" class="anchor"></a>Classification</h2>
<p>The first step in the <code>splendid</code> pipeline is classification. Given explanatory data and a reference response of classes, we wish to build a classifier that can accurately predict class representation in a separate validation data set. To avoid overfitting, we split the given data into a training set and test set. The training set is constructed by taking a random sample with replacement of all samples. The test set is comprised of all samples which are not included in the corresponding training set, also known as an OOB sample. This is a simple bootstrap resampling scheme, which we replicate a sufficient number of times to capture sampling variability.</p>
<div id="algorithms" class="section level3">
<h3 class="hasAnchor">
<a href="#algorithms" class="anchor"></a>Algorithms</h3>
<p>There exist a vast number of classification algorithms. Those currently in <code>splendid</code> are:</p>
<ul>
<li>Prediction Analysis for Microarrays (<code>"pam"</code>)</li>
<li>Support Vector Machines using a Radial-Basis Kernel(<code>"svm"</code>)</li>
<li>Random Forests (<code>"rf"</code>)</li>
<li>Linear Discriminant Analysis (<code>"lda"</code>)</li>
<li>Shrinkage Linear Discriminant Analysis (<code>"slda"</code>)<sup>1</sup>
</li>
<li>Shrinkage Diagonal Discriminant Analysis (<code>"sdda"</code>)<sup>1</sup>
</li>
<li>Multinomial Logistic Regression using
<ul>
<li>Generalized Linear Model with no penalization (<code>"mlr_glm"</code>)<sup>2</sup>
</li>
<li>GLM with LASSO penalty (<code>"mlr_lasso"</code>)<sup>2</sup>
</li>
<li>GLM with ridge penalty (<code>"mlr_ridge"</code>)<sup>2</sup>
</li>
<li>Neural Networks (<code>"mlr_nnet"</code>)</li>
</ul>
</li>
<li>Neural Networks with One Hidden Layer (<code>"nnet"</code>)</li>
<li>Naive Bayes (<code>"nbayes"</code>)</li>
<li>Adaptive Boosting using Decision Trees (<code>"adaboost"</code>)</li>
<li>Extreme Gradient Boosting using Decision Trees (<code>"xgboost"</code>)</li>
<li>K-Nearest Neighbours (<code>"knn"</code>)</li>
</ul>
<p>These algorithms are implemented in <code><a href="../reference/classification.html">classification()</a></code>.</p>
</div>
<div id="hyperparameters-and-other-details" class="section level3">
<h3 class="hasAnchor">
<a href="#hyperparameters-and-other-details" class="anchor"></a>Hyperparameters and other Details</h3>
<p>Certain functions have hyperparameters that need to be tuned in order to select the best model before prediction. We use a grid search on a pre-specified range of the hyperparameters and choose the optimal values using <code>e1071::tune()</code>.</p>
<p>The ranges for the tuning parameters are:</p>
<ul>
<li>
<code>nnet</code>
<ul>
<li>
<code>size</code>: 1, 2, 3, 4, 5</li>
<li>
<code>decay</code>: 0, 0.125, 0.25, 0.375, 0.5</li>
</ul>
</li>
<li>
<code>svm</code>
<ul>
<li>
<code>gamma</code>: <code>1 / ncol(data) * 2 ^ (0:4)</code>
</li>
<li>
<code>cost</code>: 1, 2, 4, 8, 16</li>
</ul>
</li>
</ul>
<p>Some algorithms have certain properties that require data manipulations before classification.</p>
<ul>
<li>
<code>slda</code> and <code>sdda</code>: These shrinkage discriminant analysis algorithms estimate the regularization parameters analytically without having to solve using resampling methods.</li>
<li>
<code>multinom_nnet</code> and <code>nnet</code>: We increase the maximum number of allowable weights to a sufficiently large number to avoid ill-fitted models.</li>
<li>
<code>knn</code>: Since K-Nearest Neighbours is a non-parametric method, there <em>is</em> no model object upon which predictions are made, and predicted classes are directly found. To obtain prediction probabilities by class, we use a distance matrix between the training and test set samples.</li>
<li>
<code>pam</code>: a uniform prior is used in the training step.</li>
<li>
<code>xgboost</code>: the objective function is soft probability, and the evaluation metric is log loss.</li>
</ul>
</div>
<div id="feature-selection" class="section level3">
<h3 class="hasAnchor">
<a href="#feature-selection" class="anchor"></a>Feature Selection</h3>
<p>We use Recursive Feature Elimination (RFE)<sup>3</sup> on <code>lda</code>, <code>rf</code>, and <code>svm</code> to reduce the dimensionality before classification. In <code>svm</code>, we do this before tuning because of the computational complexity. However, we can embed this feature selection within the tuning step for the other three algorithms. Set <code>rfe = TRUE</code> in <code>splendid</code> to use feature selection.</p>
<p>One limitation of RFE is that an a priori set of feature subset sizes need to be specified, determining the search space for the algorithm. By default, we set the <code>sizes</code> parameter to be every 5 integers from 0 up to one-half of the smallest class size. Recall the class sizes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(class)
<span class="co">#&gt; class</span>
<span class="co">#&gt; DIF.C4 IMM.C2 MES.C1 PRO.C5 </span>
<span class="co">#&gt;    135    107    109    138</span></code></pre></div>
<p>We tell RFE to search for the best models with 5, 10, …, 50 features. Cross-validation with 2 folds is used in the algorithm.</p>
</div>
</div>
<div id="prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#prediction" class="anchor"></a>Prediction</h2>
<p>There is a different prediction method for each classifier, since they are all imported from an external package. The <code><a href="../reference/prediction.html">prediction()</a></code> function calls each method based on the class of the model output from <code><a href="../reference/classification.html">classification()</a></code>. <code><a href="../reference/prediction.html">prediction()</a></code> also performs some manipulations so that the results all have the same data structure: unnamed factors, with labels given in the same order as the true class labels.</p>
<p>The method for <code>"pam"</code> is an exception: the output from <code><a href="../reference/prediction.html">prediction()</a></code> not only has the predicted class labels, but also the cross-validated threshold value calculated from the training set to use in prediction on the test set, named <code>delta</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(sl_result$preds, <span class="dt">max.level =</span> <span class="dv">2</span>, <span class="dt">list.len =</span> <span class="dv">2</span>)
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ slda:List of 2</span>
<span class="co">#&gt;   ..$ : Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 1 3 3 4 2 4 2 4 2 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "prob")= num [1:189, 1:4] 0.0202 1 0 0 0 ...</span>
<span class="co">#&gt;   .. .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..- attr(*, "class.true")= Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 1 3 3 3 2 4 2 4 2 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "class.thres")= Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 1 3 3 4 2 4 2 4 2 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "class.prop")= num 1</span>
<span class="co">#&gt;   ..$ : Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 4 4 4 1 1 1 1 2 1 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "prob")= num [1:170, 1:4] 0.000858 0.039292 0 0 0.999683 ...</span>
<span class="co">#&gt;   .. .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..- attr(*, "class.true")= Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 3 4 4 4 1 1 4 2 1 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "class.thres")= Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 4 4 4 1 1 1 1 2 1 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "class.prop")= num 1</span>
<span class="co">#&gt;  $ knn :List of 2</span>
<span class="co">#&gt;   ..$ : Factor w/ 4 levels "DIF.C4","IMM.C2",..: 4 1 3 3 3 2 4 2 4 2 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "prob")= num [1:189, 1:4] 0 1 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   .. .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..- attr(*, "class.true")= Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 1 3 3 3 2 4 2 4 2 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "class.thres")= Factor w/ 5 levels "DIF.C4","IMM.C2",..: 4 1 3 3 3 2 4 2 4 2 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "class.prop")= num 0.963</span>
<span class="co">#&gt;   ..$ : Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 3 4 4 1 1 2 4 1 2 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "prob")= num [1:170, 1:4] 0.2 0.2 0 0 1 1 0.4 0 0.6 0.4 ...</span>
<span class="co">#&gt;   .. .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..- attr(*, "class.true")= Factor w/ 4 levels "DIF.C4","IMM.C2",..: 3 3 4 4 4 1 1 4 2 1 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "class.thres")= Factor w/ 5 levels "DIF.C4","IMM.C2",..: 3 3 4 4 1 1 2 4 1 2 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "class.prop")= num 0.988</span>
<span class="co">#&gt;   [list output truncated]</span></code></pre></div>
<div id="attributes" class="section level3">
<h3 class="hasAnchor">
<a href="#attributes" class="anchor"></a>Attributes</h3>
<p>Note that each predicted class has an attribute that shows the class probabilities in a matrix. We ensure that the class probabilities for every sample sum to one by making a small adjustment to one of the classes depending on whether the probability was over or under one. These matrices are useful to compute evaluation measures and generate discriminating graphs that we detail below, and is stored in <code>attr(*, "prob")</code>.</p>
<p>To obtain better performance under evaluation metrics, we may want to exclude samples which have a maximum class probability below a threshold. For example, we may only compare the true test labels with the corresponding OOB samples where the winning class has a probability at least 50%. Samples below this threshold are labelled as <code>"unclassified"</code>. If the threshold results in all samples being unclassified, then we use the unfiltered, original predicted labels for evaluation. This object is stored in <code>attr(*, "class.thres")</code>.</p>
<p>The final attribute of a prediction object is stored <code>attr(*, "class.prop")</code>, showing the proportion of classified cases. This value is inversely proportional to the <code>threshold</code> setting.</p>
</div>
</div>
<div id="evaluation" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluation" class="anchor"></a>Evaluation</h2>
<div id="metrics" class="section level3">
<h3 class="hasAnchor">
<a href="#metrics" class="anchor"></a>Metrics</h3>
<p>Evaluation measures are important because they tell us the prediction performance of a classifier. The table below shows all the measures for <code>svm</code>, for two bootstrap replicates. The <code>logloss</code> and <code>auc</code> measures make use of the prediction probabilities, whereas the rest are computed from the multiclass confusion matrix of reference and predicted class labels.</p>
<table class="table">
<caption>SVM Evaluation Metrics</caption>
<thead><tr class="header">
<th></th>
<th align="right">1</th>
<th align="right">2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>logloss</td>
<td align="right">0.3050324</td>
<td align="right">0.2500494</td>
</tr>
<tr class="even">
<td>auc</td>
<td align="right">0.9840834</td>
<td align="right">0.9901929</td>
</tr>
<tr class="odd">
<td>accuracy</td>
<td align="right">0.8938547</td>
<td align="right">0.9119497</td>
</tr>
<tr class="even">
<td>macro_ppv</td>
<td align="right">0.8939579</td>
<td align="right">0.9109104</td>
</tr>
<tr class="odd">
<td>macro_sensitivity</td>
<td align="right">0.8953092</td>
<td align="right">0.9175498</td>
</tr>
<tr class="even">
<td>macro_f1</td>
<td align="right">0.8939720</td>
<td align="right">0.9122388</td>
</tr>
<tr class="odd">
<td>mcc</td>
<td align="right">0.8580875</td>
<td align="right">0.8833621</td>
</tr>
<tr class="even">
<td>micro_mcc</td>
<td align="right">0.8584730</td>
<td align="right">0.8825996</td>
</tr>
<tr class="odd">
<td>ppv.DIF.C4</td>
<td align="right">0.8703704</td>
<td align="right">0.9487179</td>
</tr>
<tr class="even">
<td>ppv.IMM.C2</td>
<td align="right">0.8461538</td>
<td align="right">0.8611111</td>
</tr>
<tr class="odd">
<td>ppv.MES.C1</td>
<td align="right">0.9545455</td>
<td align="right">0.9189189</td>
</tr>
<tr class="even">
<td>ppv.PRO.C5</td>
<td align="right">0.9047619</td>
<td align="right">0.9148936</td>
</tr>
<tr class="odd">
<td>sensitivity.DIF.C4</td>
<td align="right">0.8867925</td>
<td align="right">0.8222222</td>
</tr>
<tr class="even">
<td>sensitivity.IMM.C2</td>
<td align="right">0.9166667</td>
<td align="right">0.9687500</td>
</tr>
<tr class="odd">
<td>sensitivity.MES.C1</td>
<td align="right">0.9333333</td>
<td align="right">0.9444444</td>
</tr>
<tr class="even">
<td>sensitivity.PRO.C5</td>
<td align="right">0.8444444</td>
<td align="right">0.9347826</td>
</tr>
<tr class="odd">
<td>f1.DIF.C4</td>
<td align="right">0.8785047</td>
<td align="right">0.8809524</td>
</tr>
<tr class="even">
<td>f1.IMM.C2</td>
<td align="right">0.8800000</td>
<td align="right">0.9117647</td>
</tr>
<tr class="odd">
<td>f1.MES.C1</td>
<td align="right">0.9438202</td>
<td align="right">0.9315068</td>
</tr>
<tr class="even">
<td>f1.PRO.C5</td>
<td align="right">0.8735632</td>
<td align="right">0.9247312</td>
</tr>
<tr class="odd">
<td>mcc.DIF.C4</td>
<td align="right">0.8267917</td>
<td align="right">0.8424771</td>
</tr>
<tr class="even">
<td>mcc.IMM.C2</td>
<td align="right">0.8493493</td>
<td align="right">0.8903614</td>
</tr>
<tr class="odd">
<td>mcc.MES.C1</td>
<td align="right">0.9253394</td>
<td align="right">0.9112477</td>
</tr>
<tr class="even">
<td>mcc.PRO.C5</td>
<td align="right">0.8339015</td>
<td align="right">0.8937284</td>
</tr>
</tbody>
</table>
<ul>
<li>
<code>logloss</code>: Multiclass log loss should be minimized</li>
<li>
<code>auc</code>: Area Under the Curve compares trade-off between sensitivity and specificity</li>
<li>
<code>accuracy</code>: number of correctly classified samples out of all samples</li>
<li>
<code>ppv</code>: also known as precision</li>
<li>
<code>sensitivity</code>: also known as recall</li>
<li>
<code>f1</code>: balanced metric between precision and recall</li>
<li>
<code>mcc</code>: Matthew’s correlation coefficient</li>
</ul>
<p>Note that in the multiclass case, we have variants for some of the measures listed above, depending on how we are looking at the data. A <code>macro</code>-averaged metric calculates said metric on each of the one-vs-all confusion matrices and then takes the mean. A <code>micro</code>-averaged metric is calculated on the element-wise sum of all one-vs-all confusion matrices.</p>
<p>Hence, there are <code>macro_ppv</code>, <code>macro_sensitivity</code>, and <code>macro_f1</code> macro-averaged metrics, and <code>micro_mcc</code> as the “only” micro-averaged metric. It turns out that <code>accuracy</code> is mathematically equivalent to micro-averaged PPV, sensitivity, and F1-score, so we don’t redundantly add those to the list. Furthermore, calculating <code>mcc</code> in a macro-averaged way is not recommended, so we use it on the full confusion matrix.</p>
<p>Class-specific measures fill in the rest of the matrix, labelled with the naming scheme <code>metric.class</code>.</p>
</div>
<div id="plots" class="section level3">
<h3 class="hasAnchor">
<a href="#plots" class="anchor"></a>Plots</h3>
<p>To assess the performance of a classifier, we can look at a discriminating plot and reliability plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(hgsc)
class &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">attr</span>(hgsc, <span class="st">"class.true"</span>))
<span class="kw">set.seed</span>(<span class="dv">1</span>)
training.id &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">seq_along</span>(class), <span class="dt">replace =</span> <span class="ot">TRUE</span>)
test.id &lt;-<span class="st"> </span><span class="kw">which</span>(!<span class="kw">seq_along</span>(class) %in%<span class="st"> </span>training.id)
mod &lt;-<span class="st"> </span><span class="kw"><a href="../reference/classification.html">classification</a></span>(hgsc[training.id, ], class[training.id], <span class="st">"adaboost"</span>)
pred &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prediction.html">prediction</a></span>(mod, hgsc, test.id, <span class="dt">class =</span> class)
evals &lt;-<span class="st"> </span><span class="kw"><a href="../reference/evaluation.html">evaluation</a></span>(class[test.id], pred, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="ensemble_classification_files/figure-html/plots-1.png" width="576"><img src="ensemble_classification_files/figure-html/plots-2.png" width="576"></p>
</div>
</div>
<div id="ensemble-construction" class="section level2">
<h2 class="hasAnchor">
<a href="#ensemble-construction" class="anchor"></a>Ensemble Construction</h2>
<p>For each bootstrap replicate training set, we want to find the top performing classifier. We use Rank Aggregation with the Genetic Algorithm to choose the top algorithm by comparing across evaluation metrics<sup>4</sup>. In the case of log loss, we need to first invert its value since it its objective function is minimization. Class-specific measures are not included in the rank aggregation because they are interdependent. For example, a sample with a true class of “A” should not be treated differently depending on whether it was misclassified into “B” or “C”.</p>
<p>After obtaining the list of top classifiers for each bootstrap replicate, we sort them by decreasing frequency and keep the top 3 to use in the ensemble. The <code>ensemble_mods</code> output of <code>splendid</code> is a list of models fit on the full data for the top classifiers.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ol style="list-style-type: decimal">
<li>Ahdesmäki, Miika, and Korbinian Strimmer. “Feature selection in omics prediction problems using cat scores and false nondiscovery rate control.” <em>The Annals of Applied Statistics</em> 4.1 (2010): 503-519.</li>
<li><a href="https://cran.r-project.org/web/packages/glmnet/vignettes/glmnet_beta.pdf" class="uri">https://cran.r-project.org/web/packages/glmnet/vignettes/glmnet_beta.pdf</a></li>
<li><a href="https://topepo.github.io/caret/recursive-feature-elimination.html" class="uri">https://topepo.github.io/caret/recursive-feature-elimination.html</a></li>
<li>Pihur, Vasyl, Susmita Datta, and Somnath Datta. “RankAggreg, an R package for weighted rank aggregation.” <em>BMC bioinformatics</em> 10.1 (2009): 62.</li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#overview">Overview</a></li>
      <li><a href="#classification">Classification</a></li>
      <li><a href="#prediction">Prediction</a></li>
      <li><a href="#evaluation">Evaluation</a></li>
      <li><a href="#ensemble-construction">Ensemble Construction</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Derek Chiu, Aline Talhouk, Dustin Johnson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
